backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/disk/image/ConvertDiskCommand.java

```diff
@@ -0,0 +1,366 @@
+
+    public ConvertDiskCommand(T parameters, CommandContext cmdContext) {
+        super(parameters, cmdContext);
+    }
+
+    public ConvertDiskCommand(Guid commandId) {
+        super(commandId);
+    }
+
+    @Override
+    public void init() {
+        super.init();
+        if (getDiskImage() == null) {
+            DiskImage diskImage = diskImageDao.getSnapshotById(getLocationInfo().getImageId());
+            setDiskImage(diskImage);
+        }
+    }
+
+    @Override
+    protected boolean validate() {
+        StorageDomain storageDomain = storageDomainDao.get(getDiskImage().getStorageIds().get(0));
+        if (storageDomain == null) {
+            return failValidation(EngineMessage.STORAGE_DOMAIN_DOES_NOT_EXIST);
+        }
+
+        // We do not support chains currently
+        if (diskImageDao.getAllSnapshotsForImageGroup(getDiskImage().getId()).stream().count() > 1) {
+            return failValidation(EngineMessage.ACTION_TYPE_FAILED_DISK_WITH_CHAIN);
+        }
+
+        StorageDomainValidator storageDomainValidator = new StorageDomainValidator(storageDomain);
+        if (!validate(storageDomainValidator.hasSpaceForNewDisk(getDiskImage()))) {
+            return false;
+        }
+
+        if (getDiskImage().getImageStatus() == ImageStatus.LOCKED) {
+            return failValidation(EngineMessage.ACTION_TYPE_FAILED_DISK_IS_LOCKED);
+        }
+
+        // We do not support live stuff currently
+        Map<Boolean, List<VM>> vms = vmDao.getForDisk(getDiskImage().getId(), true);
+
+        boolean runningVms = vms.computeIfAbsent(Boolean.TRUE, b -> Collections.emptyList())
+                .stream()
+                .anyMatch(vm -> vm.isRunningOrPaused());
+        if (runningVms) {
+            return failValidation(EngineMessage.ACTION_TYPE_FAILED_VM_IS_RUNNING);
+        }
+
+        // Make sure unsupported combinations aren't used (sparse-raw on block, etc)
+        return ImagesHandler.checkImageConfiguration(storageDomain.getStorageStaticData(),
+                getVolumeType(getDiskImage()),
+                getVolumeFormat(getDiskImage()),
+                getDiskImage().getBackup(),
+                getReturnValue().getValidationMessages());
+    }
+
+    @Override
+    protected void executeCommand() {
+        Guid hostForExecution = vdsCommandsHelper.getHostForExecution(getDiskImage().getStoragePoolId(),
+                FeatureSupported::isConvertDiskSupported);
+        if (Guid.isNullOrEmpty(hostForExecution)) {
+            log.error("Could not find a host to perform conversion");
+            return;
+        }
+
+        getParameters().setVdsRunningOn(hostForExecution);
+
+        // Disable pre-polling nonsense
+        getParameters().setEntityInfo(new EntityInfo(VdcObjectType.Disk, getDiskImage().getId()));
+        getParameters().setNewVolGuid(Guid.newGuid());
+
+        imagesHandler.updateImageStatus(getDiskImage().getImageId(), ImageStatus.LOCKED);
```

Comment:
missing compensation here in case of a failure later on (or engine restart)?

Reply:
added